File: src/Coachable/sorting/QuickSort.java
package Coachable.sorting;


import java.util.Arrays;


public class QuickSort {

	public static void main(String[] args) {

		//int[] arr = new int[]{50, 70, 60, 90, 40, 80, 10, 20, 30, 9999};
		//int[] arr = new int[]{69, 65, 83, 89, 81, 85, 69, 83, 84, 73, 79, 78, 9999};
		int[] arr = new int[]{69, 65, 83, 89, 81, 85, 69, 83, 84, 73, 79, 78};
		System.out.println(Arrays.toString(arr));


		QuickSorting3Way(arr, arr.length);
		System.out.println(Arrays.toString(arr));

		//QuickSorting(arr, 0, arr.length - 1);
		//System.out.println(Arrays.toString(arr));
	}



	// quick sorts only 3 distant elements
	private static void QuickSorting3Way(int[] arr, int size) {

		int low = 0;
		int high = size - 1;
		int mid = 0;

		while (mid <= high) {
			if (arr[mid] == 0) {
				int temp = arr[low];
				arr[low] = arr[mid];
				arr[mid] = temp;
			} else if (arr[mid] == 1)
				mid++;
			else {
				int temp = arr[high];
				arr[high] = arr[mid];
				arr[mid] = temp;
				high--;
			}
		}
	}

	private static void QuickSorting(int[] arr, int low, int high) {

		// recursively call quick sorting
		if (low < high) {

			// get the partition using the entire array
			int partitionIdx = partition(arr, low, high);


			/**
			 * when using the while loop partition, exclude the right most element
			 * since that is the pivot, and it's used as right most marker
			 * recursive call on the left half
			 */
			//QuickSorting(arr, low, partitionIdx);

			// recursive all on the left half
			QuickSorting(arr, low, partitionIdx - 1);

			// recursive call on the right half
			QuickSorting(arr, partitionIdx + 1, high);
		}
	}


	private static int partition(int[] arr, int low, int high) {
		// assume last element is the pivot
		int pivot = arr[high];

		// pointer 'i' tracks the latest element that was smaller than pivot
		int i = low - 1;
		// pointer 'j' goes from low to the high-1, since pivot is at high
		for (int j = low; j < high; j++) {

			// search for item that is <= pivot
			if (arr[j] <= pivot) {

				// move 'i', since i current points to the element that is smaller than pivot
				i++;

				// swap the elements at i and j
				int temp = arr[j];
				arr[j] = arr[i];
				arr[i] = temp;
			}
		}

		// move 'i', swap that element with pivot, and return i
		i++;
		arr[high] = arr[i];
		arr[i] = pivot;

		return i;
	}

	private static int partitionUsingWhileLoop(int[] arr, int l, int h) {

		int pivot = arr[l];

		int i = l;
		int j = h;

		do {
			// find item > pivot, using i
			do i++;
			while (arr[i] <= pivot);

			// find item <= pivot, using j
			do j--;
			while (arr[j] > pivot);

			// swap both items, only if pointers didn't cross
			if (i < j) {
				int temp = arr[j];
				arr[j] = arr[i];
				arr[i] = temp;
			}
		} while (i < j);

		// bring the pivot to j
		arr[l] = arr[j];
		arr[j] = pivot;

		return j;
	}

}

File: src/Coachable/1-DSA/4-String,Sorting/QuickSort.py
"""
QuickSort.py
"""

"""
Quick Sort
Stable/unstable sorting
Time: O()
Space: O()

"""


def quick_sort(array: list[int]) -> None:
    sort(array, 0, len(array) - 1)


def sort(array, low, high):
    print("after partition")
    print(array)
    if low >= high:
        return
    partition_idx = find_partition(array, low, high)
    sort(array, low, partition_idx - 1)
    sort(array, partition_idx + 1, high)


def find_partition(array, low, high):
    pivot = array[low]
    i, j = low + 1, high

    while True:

        # 'i' goes from left to right, to find element > pivot
        while array[i] <= pivot and i < high:
            i += 1
            # if i == high:
            #     break

        # 'j' goes from right to left, to find element <= pivot
        while array[j] > pivot and j > low:
            j -= 1
            # if j == low:
            #     break

        if i >= j:
            break

        # swap the elements
        array[i], array[j] = array[j], array[i]

    # swap pivot with j
    array[low], array[j] = array[j], array[low]

    return j


# arr = [18, 0, 18, -1, -5, 10, 2, 1]
arr = [3, 8, 2, 15, 27, 21, 17, 10, 16, 7, 24, 0, 4, 6, 18, 5]
quick_sort(arr)
print(arr)
